### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by 4ertya**

### CODE REVIEW

File: api/src/main/java/com/epam/esm/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/orders")
+public class OrderController {
+    private final OrderService orderService;
+    private final HateoasBuilder hateoasBuilder;
+
+
+    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE)
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/OrderController.java]: (consumes = MediaType.APPLICATION_JSON_VALUE) -why do you still use this?

File: api/src/main/java/com/epam/esm/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    @ResponseStatus(HttpStatus.CREATED)
+    public OrderDto createOrder(@RequestBody NewOrderDto newOrderDto) {
+        OrderDto orderDTO = orderService.addOrder(newOrderDto);
+        return hateoasBuilder.addLinksForOrder(orderDTO);
+    }
+
+
+    @GetMapping
+    public RepresentationModel<?> getAllOrders(@RequestParam Map<String, String> params) {
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/controller/OrderController.java]: ? - it does not look nice...

File: api/src/main/java/com/epam/esm/util/PaginationPreparer.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+
+@Component
+public class PaginationPreparer {
+
+    public List<Link> preparePaginationLinks(Object invocationValue, Map<String, String> params, long entityCount) {
+        int currentPage = getPageNumber(params);
+        long pageCount = calculatePageCount(params, entityCount);
+        List<Link> links = new ArrayList<>();
+        if (pageCount != 1 && currentPage != 1) {
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/util/PaginationPreparer.java]: 1 - magic numbers

File: api/src/main/java/com/epam/esm/util/PaginationPreparer.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;
+
+@Component
+public class PaginationPreparer {
+
+    public List<Link> preparePaginationLinks(Object invocationValue, Map<String, String> params, long entityCount) {
```
Comment (DmitrySay) [inline api/src/main/java/com/epam/esm/util/PaginationPreparer.java:16]: Map<String, String>  - better to use a class which you can validate above

File: core/src/main/java/com/epam/esm/dto/TagDto.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import org.springframework.hateoas.server.core.Relation;
+@EqualsAndHashCode(callSuper = true)
+@Data
+@NoArgsConstructor
+@Relation(itemRelation = "tag", collectionRelation = "tags")
+public class TagDto extends RepresentationModel<TagDto> {
+    private Long id;
+    private String name;
+
+    public Long getId() {
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/dto/TagDto.java]: lombok?

File: core/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    private CriteriaQuery<Certificate> buildCriteriaQuery(List<Specification> specifications) {
+        CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Certificate> criteriaQuery = criteriaBuilder.createQuery(Certificate.class);
+        Root<Certificate> root = criteriaQuery.from(Certificate.class);
+        criteriaQuery.orderBy(criteriaBuilder.asc(root.get(ID)));
+        List<Predicate> predicateList = new ArrayList<>();
+        specifications.forEach(specification -> {
+            if (specification instanceof SearchSpecification)
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: specification instanceof SearchSpecification ?

File: core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    @Override
+    public List<CertificateDto> findCertificates(Map<String, String> params) {
+        basicValidator.checkOrderByParams(params);
+        basicValidator.checkTagNameParams(params);
+        paginationValidator.validatePaginationParams(params);
+        List<Specification> specifications = certificateSpecificationCreator.generateQueryCriteria(params);
+        long elementsCount = certificateRepository.getCount(specifications);
+        paginationValidator.validatePageNumber(params, elementsCount);
+        List<CertificateDto> certificates = new ArrayList<>();
+        int limit = Integer.parseInt(params.get("size"));
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/CertificateServiceImpl.java]: Integer.parseInt(params.get("size"))

File: core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    private final TagMapper tagMapper;
+    private final PaginationValidator paginationValidator;
+    private final EntityValidator entityValidator;
+    private final BasicValidator basicValidator;
+
+    @Override
+    public List<TagDto> findAllTags(Map<String, String> params) {
+        paginationValidator.validatePaginationParams(params);
+        int limit = Integer.parseInt(params.get("size"));
+        int offset = (Integer.parseInt(params.get("page")) - 1) * limit;
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/service/impl/TagServiceImpl.java]: (Integer.parseInt(params.get("page")) - 1)

File: core/src/main/java/com/epam/esm/validation/BasicValidator.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    private String validateTagNames(String tagNamesAsString) {
+        List<String> validatedTags = new ArrayList<>();
+        String[] tagNames = tagNamesAsString.split(Constants.COMMA);
+        for (String tagName : tagNames) {
+            validateStringField(tagName, Constants.TAG_NAME);
+            validatedTags.add(tagName.trim().toLowerCase());
+        }
+        return String.join(Constants.COMMA, validatedTags.toArray(new String[0]));
```
Comment (DmitrySay) [inline core/src/main/java/com/epam/esm/validation/BasicValidator.java]: toArray(new String[0]));

File: core/src/main/resources/bd.migrate/schema.sql [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+        primary key (id)
+);
+
+create table tags
+(
+    id   bigserial not null,
+    name varchar(255),
+    constraint tags_pkey
+        primary key (id),
+    constraint uk_t48xdq560gs3gap9g7jg36kgc
```
Comment (DmitrySay) [inline core/src/main/resources/bd.migrate/schema.sql]: constraint uk_t48xdq560gs3gap9g7jg36kgc

File: core/src/main/resources/bd.migrate/schema.sql [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+    user_id    bigint,
+    constraint orders_pkey
+        primary key (id)
+);
+
+create table orders_certificates
+(
+    orders_id       bigint not null,
+    certificates_id bigint not null,
+    constraint fk80siohi4i1dhggs033y44gh97
```
Comment (DmitrySay) [inline core/src/main/resources/bd.migrate/schema.sql]: constraint fk80siohi4i1dhggs033y44gh97

File: core/src/main/resources/bd.migrate/schema.sql [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+
+create table certificates_tags
+(
+    certificates_id bigint not null,
+    tags_id         bigint not null,
+    constraint certificates_tags_pkey
+        primary key (certificates_id, tags_id),
+    constraint fk7rhfn769tsnnq5e0mbtnglh6w
+        foreign key (certificates_id) references certificates,
+    constraint fks9jnjgsosxmo7mrpqyipfag78
```
Comment (DmitrySay) [inline core/src/main/resources/bd.migrate/schema.sql]: constraint fks9jnjgsosxmo7mrpqyipfag78

File: core/src/main/resources/bd.migrate/schema.sql [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+    revtype          smallint,
+    create_date      timestamp with time zone,
+    description      varchar(255),
+    duration         integer,
+    last_update_date timestamp with time zone,
+    name             varchar(255),
+    price            numeric(19, 2),
+    constraint certificates_audit_log_pkey
+        primary key (id, rev),
+    constraint fkqwbymh7o15dgqmsok0ssh854k
```
Comment (DmitrySay) [inline core/src/main/resources/bd.migrate/schema.sql]: fkqwbymh7o15dgqmsok0ssh854k

File: core/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java (hunk 6 above)
Comment (4ertya) [reply core/src/main/java/com/epam/esm/repository/impl/CertificateRepositoryImpl.java]: Thanks for the comment, but I don't understand what you mean.

**Pull request merged**